# Using Tensorflow Dataset and Modeling


# Imports

In [6]:
from google.colab import drive
drive.mount('/gdrive')

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
# !pip uninstall tf-keras
# !pip install tensorflow==2.16.1

In [1]:
import keras
import tensorflow as tf
print("Keras Current Version:", keras.__version__, "Tensorflow Current Version:", tf.__version__)

Keras Current Version: 3.13.2 Tensorflow Current Version: 2.20.0


In [2]:
import pandas as pd
# burada bir yerde MinMaxScaler'ı import ediniz
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Dense
import tensorflow as tf

# Data Preparation

## Task 1: proprocess_data Fonksiyonunda MinMaxScaler'ı Kullanınız.

- Import bölümünde MinMaxScaler'ı import etmeyi unutmayınız.
- preprocess_data fonksiyonu içinde scaler olarak MinMaxScaler'ı kullanınız.
- Sonrasınra train validasyon ayrımını yapınız.

In [7]:
def preprocess_data(filepath):
    data = pd.read_csv(filepath)
    # Buraya MinMaxScaler ekleyiniz.
    scaler = MinMaxScaler()
    X = scaler.fit_transform(data.drop('Outcome', axis=1))
    y = data['Outcome'].values
    return X, y


X, y = preprocess_data('/gdrive/MyDrive/Colab Notebooks/03_neural_network_course_materials/data_sets/diabetes.csv')

# train validasyon ayrımını yapınız.

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

print("X_train shape:", X_train.shape)

X_train shape: (614, 8)


# Create Tensorflow Dataset


## Task 2: Train Seti ve Validasyon Seti için Tensorflow Dataset Oluşturunuz

In [10]:
train_dataset = tf.data.Dataset.from_tensor_slices((X_train, y_train))
val_dataset = tf.data.Dataset.from_tensor_slices((X_val,y_val))

## Task 3: val_dataset'in 1 Gözlemini İnceleyiniz

In [12]:
for features, label in val_dataset.take(1):
  print('Features:', features.numpy(), 'Label:', label.numpy())

Features: [0.35294118 0.49246231 0.47540984 0.33333333 0.22458629 0.50670641
 0.15029889 0.36666667] Label: 0


## Task 4: buffer_size'ı tüm veri seti boyutu olarak ve batch'ı 32 olacak şekilde train ve validasyon setlerini biçimlendiriniz. Validasyon seti için sadece batch uyguladığımızı unutmayın.

In [17]:
train_dataset = train_dataset.shuffle(buffer_size=len(train_dataset)).batch(batch_size=32)
val_dataset = val_dataset.batch(batch_size=32)

## Task 5: Validasyon Seti için 1 Gözlemi İnceleyiniz. 10. batch'in feature değerleri nelerdir? Label değeri nedir? Amelasyon ile tek tek sayarak inceleyiniz.

In [19]:
for features, labels in val_dataset.take(1):
    print('Features:', features.numpy(), 'Label:', labels.numpy())

Features: [[0.35294118 0.49246231 0.47540984 0.33333333 0.22458629 0.50670641
  0.15029889 0.36666667]
 [0.11764706 0.56281407 0.6147541  0.32323232 0.         0.53204173
  0.02988898 0.        ]
 [0.11764706 0.54271357 0.52459016 0.         0.         0.45901639
  0.03415884 0.        ]
 [0.47058824 0.53768844 0.6557377  0.         0.         0.36661699
  0.33219471 0.21666667]
 [0.41176471 0.68341709 0.73770492 0.         0.         0.44560358
  0.05636208 0.48333333]
 [0.35294118 0.51758794 0.59016393 0.32323232 0.22458629 0.56184799
  0.10503843 0.56666667]
 [0.05882353 0.35678392 0.39344262 0.18181818 0.08983452 0.30402385
  0.10461144 0.01666667]
 [0.         0.5879397  0.         0.         0.         0.50372578
  0.3646456  0.38333333]
 [0.23529412 0.77386935 0.59016393 0.29292929 0.14893617 0.46646796
  0.11101623 0.26666667]
 [0.29411765 0.73869347 0.63934426 0.         0.         0.50223547
  0.05977797 0.73333333]
 [0.58823529 0.55778894 0.57377049 0.27272727 0.         0.4

Feature :[0.29411765 0.73869347 0.63934426 0.         0.         0.50223547
  0.05977797 0.73333333]
Label : 0

# Model

## Task 6: Dersteki modeli 100 epoch sayısı ile tekrar eğitiniz ve loss ve accuracy değerlerini yorumlayınız. Tüm işlemleri tek bir hücrede yapınız.

In [23]:
model = Sequential([
    Input(shape=(X_train.shape[1],)),
    Dense(32, activation='relu'),
    Dense(16, activation='relu'),
    Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(train_dataset,
                    epochs=100,
                    validation_data=val_dataset)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 3s 74ms/step - accuracy: 0.6531 - loss: 0.6692 - val_accuracy: 0.6429 - val_loss: 0.6648
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6531 - loss: 0.6585 - val_accuracy: 0.6429 - val_loss: 0.6594
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6531 - loss: 0.6526 - val_accuracy: 0.6429 - val_loss: 0.6541
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6531 - loss: 0.6461 - val_accuracy: 0.6429 - val_loss: 0.6468
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6531 - loss: 0.6390 - val_accuracy: 0.6429 - val_loss: 0.6409
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6531 - loss: 0.6337 - val_accuracy: 0.6429 - val_loss: 0.6360
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6515 - loss: 0.6287 - val_accuracy: 0.6429 - val_loss: 0.6301
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6531 - loss: 0.6230 - val_accuracy: 0.6494 - 

In [ ]:
# derste elde edilen değerler
# val_loss: 0.59
# val_accuracy: 0.70

In [ ]:
# Task 6 sonucu:
# Validation Loss: 0.52
# Validation Accuracy: 0.73
# Yorum: Loss azalirken Accuracy artmis.Bu durum gelisme gibi görünsede, önceki epochlarda val_accuracy artip sonra tekrar bir azalis trendine giriyor ve val_loss ilk 50 epoch sonrasinda belirli bir aralikta seyrediyor. Bu durum modelin  overfit oldugunu gösterebilir. Egitimi biraz daha kisa tutmak iyi olabilirdi.

## Task 7: 5 katmanlı ve nöron sayıları sırasıyla 32, 64, 128, 256, 512 olan bir model kurunuz ve sonuçları değerlendiriniz. Diğer özelliklerde bir değişiklik olmamalı.

In [24]:
model = Sequential([
    Input(shape=(X_train.shape[1],)),
    Dense(32, activation='relu'),
    Dense(64, activation='relu'),
    Dense(128, activation='relu'),
    Dense(256, activation='relu'),
    Dense(512, activation='relu'),
    Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(train_dataset,
                    epochs=100,
                    validation_data=val_dataset)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 4s 92ms/step - accuracy: 0.6482 - loss: 0.6494 - val_accuracy: 0.6429 - val_loss: 0.6419
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6824 - loss: 0.6132 - val_accuracy: 0.7013 - val_loss: 0.5861
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7296 - loss: 0.5550 - val_accuracy: 0.7013 - val_loss: 0.5584
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7296 - loss: 0.5284 - val_accuracy: 0.6948 - val_loss: 0.5537
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7459 - loss: 0.5221 - val_accuracy: 0.6688 - val_loss: 0.5518
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7524 - loss: 0.5011 - val_accuracy: 0.7143 - val_loss: 0.5447
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7476 - loss: 0.4993 - val_accuracy: 0.7468 - val_loss: 0.5487
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7557 - loss: 0.4905 - val_accuracy: 0.7338 - 

In [ ]:
# Yorum: val_loss degeri giderek artti val_accuracy azalis trendini sürdüremedi. Model overfit'in ötesine gecti.
# Neden?: Ag asiri derin oldugundan bir ögrenme trendi yakalayamadi. val_loss degeri patladi azalmasi beklenirken bir noktadan sonra artis trendine girdi. Modelimiz bu problem icin asiri büyük oldu. Bu sebeple stabil bir cikti üretmek artik imkansiz hale geldi.


## Task 8: Acaba Epoch Sayısını Arttırsak bir Faydası Olur mu? 1000 ile deneyelim. Sonuçları yorumlayınız.

In [25]:
model = Sequential([
    Input(shape=(X_train.shape[1],)),
    Dense(32, activation='relu'),
    Dense(64, activation='relu'),
    Dense(128, activation='relu'),
    Dense(256, activation='relu'),
    Dense(512, activation='relu'),
    Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(train_dataset,
                    epochs=1000,
                    validation_data=val_dataset)

Epoch 1/1000
20/20 ━━━━━━━━━━━━━━━━━━━━ 4s 136ms/step - accuracy: 0.6368 - loss: 0.6609 - val_accuracy: 0.6429 - val_loss: 0.6508
Epoch 2/1000
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6482 - loss: 0.6350 - val_accuracy: 0.6753 - val_loss: 0.6175
Epoch 3/1000
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6775 - loss: 0.5943 - val_accuracy: 0.6688 - val_loss: 0.6196
Epoch 4/1000
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7150 - loss: 0.5644 - val_accuracy: 0.7273 - val_loss: 0.5617
Epoch 5/1000
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7296 - loss: 0.5433 - val_accuracy: 0.7403 - val_loss: 0.5728
Epoch 6/1000
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7215 - loss: 0.5241 - val_accuracy: 0.8052 - val_loss: 0.5217
Epoch 7/1000
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7704 - loss: 0.5028 - val_accuracy: 0.6364 - val_loss: 0.7819
Epoch 8/1000
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7541 - loss: 0.5077 - val_accuracy: 

In [26]:
# Yorum: modelimiz teorik olarak ideal accuracy'ye ulasmasina ragmen val_accuracy degeri ile arasindaki farki ve val_loss degerinin yüksekligi modelin veri setini tüm hata gürültü ve aykiri degerleri ile ezberlediginin göstergesi.